scripts/RelayHubGasEstimation.ts

```diff
@@ -0,0 +1,641 @@
+    noPayment: true,
+    noTokenGas: true,
+    noTokenContract: true,
+    noCorrection: true,
+  });
+  console.log(
+    `Destination Call with enveloping - Gas Used: ${gasUsedRound2.toNumber()}, CumulativeGasUsed: ${cumulativeGasUsedRound2}\n`
+  );
+
+  logTitle('Gas estimation scenarios');
+
+  await deployAndSetup();
+
+  const encodedFunction = recipient.interface.encodeFunctionData(
+    'emitMessage',
+    [message]
+  );
+
+  const nonceBefore = await forwarder.nonce();
+
+  const completeReq = cloneRelayRequest(relayRequest, {
+    request: {
+      data: encodedFunction,
+      nonce: nonceBefore.toString(),
+      tokenAmount: '0',
+      tokenGas: '0',
+    },
+  });
+
+  const { signature: sig } = await getSuffixDataAndSignature(
+    forwarder,
+    completeReq,
+    owner
+  );
+
+  const relayCallResult = await relayHub
+    .connect(relayWorker)
+    .relayCall(completeReq, sig, {
+      gasLimit,
+      gasPrice,
+    });
+
+  const nonceAfter = await forwarder.nonce();
+  expect(nonceBefore.add(1).toNumber()).to.equal(nonceAfter.toNumber());
+
+  const txReceipt = await relayCallResult.wait();
+
+  console.log(
+    `Gas Used: ${txReceipt.gasUsed.toString()} - Cummulative Gas Used: ${txReceipt.cumulativeGasUsed.toString()}\n`
+  );
+
+  async function forgeRequest(
+    transferReceiver: string,
+    balanceToTransfer: number,
+    fees: string,
+    token: UtilToken,
+    relayRequest: RelayRequest,
+    forwarder: SmartWallet,
+  ) {
+
+    const isSponsored = fees === '0';
+    const tokenContract = isSponsored
+      ? ethers.constants.AddressZero
+      : token.address;
+
+    const encodedFunction = token.interface.encodeFunctionData(
+      'transfer',
+      [transferReceiver, balanceToTransfer]
+    );
+
+    const estimatedDestinationCallGas = await ethers.provider.estimateGas(
+      {
+        from: forwarder.address,
+        to: token.address,
+        gasPrice,
+        data: encodedFunction,
+      }
+    );
+
+    const internalDestinationCallCost = correctEstimatedCallCost(
+      estimatedDestinationCallGas.toNumber()
+    );
+
+    const estimatedTokenPaymentGas = await ethers.provider.estimateGas({
+      from: forwarder.address,
+      to: token.address,
+      data: token.interface.encodeFunctionData('transfer', [
+        relayWorker.address,
+        fees,
+      ]),
+    });
+
+    const tokenGas = !isSponsored
+      ? {
+          tokenGas: correctEstimatedCallCost(
+            estimatedTokenPaymentGas.toNumber()
+          ),
+        }
+      : {};
+
+    const completeReq: RelayRequest = cloneRelayRequest(relayRequest, {
+      request: {
+        data: encodedFunction,
+        to: token.address,
+        nonce: (await forwarder.nonce()).toString(),
+        tokenAmount: fees,
+        tokenContract,
+        gas: internalDestinationCallCost,
+        ...tokenGas,
+      },
+    });
+
+    return completeReq;
+  }
+
+  async function estimateGasOverhead(fees: string) {
```

Comment:
I know the name is the same as before, but I'd rename it to
```suggestion
  async function estimateTokenTransferGasOverhead(fees: string) {
```

Reply:
done